src/simulator_core/solver/network/assets/solver_pipe.py

```diff
@@ -86,87 +107,268 @@ def update_loss_coefficient(self) -> None:
+
+    def _calculate_heat_loss_grid_point(self, iteration_index: int) -> float:
+        """Calculate the heat loss over a grid point."""
+        # Retrieve the internal energy
+        internal_energy_current = self._internal_energy_grid[iteration_index]
+        # Retrieve the temperature of the fluid
+        temperature_current = fluid_props.get_t(internal_energy_current)
         # Calculate the heat loss
-        self.heat_supplied = -1.0 * (
+        return -(
             self.alpha_value
-            * self.length
             * np.pi
             * self.diameter
-            * (fluid_props.get_t(internal_energy_2) - self.ambient_temperature)
+            * self.length
+            / self._grid_size
+            * (temperature_current - self.ambient_temperature)
         )
-        # self.heat_supplied = 0.0
+
+    def _internal_energy_steady_state_objective(
+        self,
+        internal_energy_iteration: float,
+        internal_energy: float,
+        mass_flow_rate: float,
+    ) -> float:
+        """Root function for the internal energy steady state equation.
+
+        :param float internal_energy_iteration: The internal energy at the current grid point.
+        :param float internal_energy: The internal energy at the previous grid point.
+        :param float mass_flow_rate: The mass flow rate of the fluid.
+
+        """
+        # Calculate the temperature of the fluid
+        temperature = fluid_props.get_t(internal_energy_iteration)
+
+        if self._use_fluid_capacity:
+            # Determine the heat transfer coefficient including fluid properties
+            heat_transfer_coefficient_fluid = self._calculate_heat_transfer_coefficient_fluid(
+                temperature=temperature,
+                mass_flow_rate=mass_flow_rate,
+            )
+            if self.alpha_value > 0.0:
+                total_heat_transfer_coefficient = 1 / (
+                    1 / heat_transfer_coefficient_fluid + 1 / self.alpha_value
+                )
+            else:
+                total_heat_transfer_coefficient = heat_transfer_coefficient_fluid
+        else:
+            total_heat_transfer_coefficient = self.alpha_value
+
+        # Return the objective function
+        element_size = self.length / self._grid_size
+        objective = mass_flow_rate * (
+            internal_energy - internal_energy_iteration
+        ) - total_heat_transfer_coefficient * np.pi * self.diameter * element_size * (
+            temperature - self.ambient_temperature
+        )
+        return float(objective)
+
+    def _update_internal_energy_grid(self) -> float:
+        # Retrieve the flow direction
+        mass_flow_rate, start_index, end_index, step = self._determine_flow_direction()
+
+        heat_supplied = 0.0
+        # Update the internal energy grid
+        for iteration_index in np.arange(start_index, end_index, step):
+            # Retrieve the previous value of the internal energy
```

Comment:
Het zou interessant zijn om te kijken of je hier iets met iterators kan doen

Reply:
Je hebt de index en step nodig voor andere stukken. Ik vermoed dat het dus lastig wordt om bijvoorbeeld al een iterator terug te geven uit _determine_flow_direction()